cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/SQLEngineCollection.java

```diff
@@ -193,6 +206,55 @@ public boolean tryStoreDirect(StageSpec stageSpec) {
     return false;
   }
 
+  @Override
+  public Set<String> tryMultiStoreDirect(PhaseSpec phaseSpec, Set<String> sinks) {
+    // Set to store names of all consumed sinks.
+    Set<String> directStoreSinks = new HashSet<>();
+
+    // Create list to store all tasks.
+    List<Future<String>> directStoreFutures = new ArrayList<>(sinks.size());
+
+    // Try to run the direct store task on all sink stages.
+    for (String sinkName : sinks) {
+      StageSpec stageSpec = phaseSpec.getPhase().getStage(sinkName);
+
+      // Check if we are able to write this output directly
+      if (stageSpec != null) {
+        // Create an async task that is used to wait for the direct store task to complete.
+        Supplier<String> task = () -> {
+          // If the direct store task succeeds, we return the sink name. Otherwise, return null.
+          if (tryStoreDirect(stageSpec)) {
+            return sinkName;
+          }
+          return null;
+        };
+        // We submit these in parallel to prevent blocking for each store task to complete in sequence.
+        Future<String> future = CompletableFuture.supplyAsync(task);
+        directStoreFutures.add(future);
+      }
+    }
+
+    // Wait for all the direct store tasks for this group, if any.
+    for (Future<String> supplier : directStoreFutures) {
+      try {
+        // Get sink name from supplier
+        String sinkName = supplier.get();
+
+        // If the sink name is not null, it means this stage was consumed successfully.
+        if (sinkName != null) {
+          directStoreSinks.add(sinkName);
+        }
+      } catch (InterruptedException e) {
+        throw Throwables.propagate(e);
+      } catch (ExecutionException e) {
+        // We don't propagate this exception as the regular sink workflow can continue.
+        LOG.warn("Execution exception when waiting for Direct Store task to complete", e);
```

Comment:
ah I see.  we don't remove it from sinks and hence regular sink workflow will continue. minor nit: maybe add that in the warn log message saying that store will continue with non-direct way. something along this line.

Reply:
Updated.